File: Algorithm/trapping_rain_water.py
class Solution:
    def trap(self,A):
        length=len(A)
        summ=0
        leftmax=0
        left=[0 for i in range(length)]
        for i in range(length):
            if leftmax<A[i]:
                leftmax=left[i]=A[i]
            else:
                left[i]=leftmax
        right=0
        print left
        for i in range(length-1,-1,-1):
            if A[i]>right:
                right=A[i]
            summ+=min(left[i],right)-A[i]
        return summ

if __name__=='__main__':
    sol=Solution()
    print sol.trap([0,1,0,2,1,0,1,3,2,1,2,1])

File: Algorithm/CPP/trapping_rain_water.cpp
# include <stdio.h>
# include <iostream>
# include <vector>
# include <string>
# include <cctype>
# include <algorithm>
# include<time.h>
using namespace std;



class Solution {
public:
	int trap(vector<int>& height) {
		int left = 0;
		int right = height.size() - 1;
		int maxLeft = 0, maxRight = 0;
		int res;
		while (left <= right)
		{
			if (height[left] <= height[right])
			{
				if (height[left]>= maxLeft) maxLeft = height[left];
				else res += maxLeft - height[left];
				left++;
			}
			else
			{
				if (height[right] >= maxRight) maxRight = height[right];
				else res += maxRight - height[right];
				right--;
			}
		}
		return res;

	}
};
